plugin/core/windows.py

```diff
@@ -310,12 +311,27 @@ def __init__(
         self._sublime = sublime
         self._handlers = handler_dispatcher
         self._restarting = False
-        self._on_closed = on_closed
         self._is_closing = False
         self._initialization_lock = threading.Lock()
         self._workspace = workspace
-        self._workspace.on_changed = self._on_project_changed
-        self._workspace.on_switched = self._on_project_switched
+        weakself = ref(self)
+
+        # A weak reference is needed, otherwise self._workspace will have a strong reference to self, meaning a
+        # cyclic dependency.
+        def on_changed(folders: List[str]) -> None:
+            this = weakself()
+            if this is not None:
+                this._on_project_changed(folders)
```

Comment:
Not sure if it would be of any use here but I've recently stumbled upon this code: https://github.com/SublimeText/PackageDev/blob/master/plugins/lib/weakmethod.py

Reply:
Interesting. WeakMethod was introduced in py3.4. I was also thinking of backporting it to py3.3, but decided that it was okay to use the code above and wait for us to switch to py3.8.